core/imageroot/usr/local/agent/bin/get-certificate

```diff
@@ -0,0 +1,125 @@
+
+        Exit codes:
+            0 - Certificate successfully obtained or updated.
+            1 - Generic error (e.g., file I/O failure, retrieval failure with existing files).
+            3 - Certificate remains unchanged.
+        """
+    )
+    parser.add_argument("--cert-file", required=True, help="Path to save the certificate file.")
+    parser.add_argument("--key-file", required=True, help="Path to save the private key file.")
+    parser.add_argument("fqdn", help="Fully qualified domain name (FQDN) for the certificate.")
+
+    args = parser.parse_args()
+    get_certificate(args.cert_file, args.key_file, args.fqdn)
+
+def get_certificate(cert_file: str, key_file: str, fqdn: str) -> None:
+    """
+    Attempts to retrieve a TLS certificate and private key. If the retrieval fails
+    and no existing certificate is found, generates a self-signed certificate.
+
+    Exit codes:
+        0 - Certificate successfully obtained or updated.
+        1 - Generic error (e.g., file I/O failure, retrieval failure with existing files).
+        3 - Certificate remains unchanged.
+
+    Args:
+        cert_file (str): Path to save the certificate.
+        key_file (str): Path to save the private key.
+        fqdn (str): Fully qualified domain name of the certificate.
+    """
+    try:
+        certificate_content, key_content = agent.get_certificate(fqdn)
+    except Exception as e:
+        print(f"Error retrieving certificate: {e}", file=sys.stderr)
+        if not os.path.exists(cert_file) or not os.path.exists(key_file):
+            print(f"Generating certificate and key for {fqdn}...")
+            certificate_content, key_content = generate_self_signed_cert(cert_file, key_file, fqdn)
+        else:
+            sys.exit(1)
+
+    existing_cert_content = read_file_content(cert_file)
+    existing_key_content = read_file_content(key_file)
+
+    if existing_cert_content == certificate_content and existing_key_content == key_content:
+        print(f"Certificate for {fqdn} is unchanged.")
+        sys.exit(3)
+
+    atomic_write(cert_file, certificate_content)
+    atomic_write(key_file, key_content)
+
+    print(f"Certificate and key for {fqdn} saved successfully.")
+    sys.exit(0)
+
+def read_file_content(file_path: str) -> str:
+    """Reads and returns the content of a file, or an empty string if the file does not exist."""
+    if os.path.exists(file_path):
+        try:
+            with open(file_path, "r") as file:
+                return file.read()
+        except IOError as e:
+            print(f"Error reading {file_path}: {e}", file=sys.stderr)
+            sys.exit(1)
+    return ""
+
+def atomic_write(file_path: str, content: str) -> None:
+    """Writes content to a temporary file and renames it to avoid corruption."""
+    temp_path = f"{file_path}.tmp"
+    try:
+        with open(temp_path, "w") as temp_file:
+            temp_file.write(content)
+        os.replace(temp_path, file_path)
```

Comment:
Only a note, the atomic operation is valid only if the temp cert file and the old cert file are in the same filesystem, otherwise the certificate will be copied.

Reply:
It can happen if the module is installed on a directory mounted from an external volume.

Reply:
Yep, I missed that.